ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1582,46 +1620,51 @@ private CompletableFuture<AppendEntriesReplyProto> appendEntriesAsync(RaftPeerId
-      }
-      followerState = updateLastRpcTime(FollowerState.UpdateType.APPEND_START);
-
-      // Check that the append entries are not inconsistent. There are 3
-      // scenarios which can result in inconsistency:
-      //      1. There is a snapshot installation in progress
-      //      2. There is an overlap between the snapshot index and the entries
-      //      3. There is a gap between the local log and the entries
-      // In any of these scenarios, we should return an INCONSISTENCY reply
-      // back to leader so that the leader can update this follower's next index.
-      final long inconsistencyReplyNextIndex = checkInconsistentAppendEntries(previous, entries);
-      if (inconsistencyReplyNextIndex > RaftLog.INVALID_LOG_INDEX) {
-        final AppendEntriesReplyProto reply = toAppendEntriesReplyProto(
-            leaderId, getMemberId(), currentTerm, followerCommit, inconsistencyReplyNextIndex,
-            AppendResult.INCONSISTENCY, callId, RaftLog.INVALID_LOG_INDEX, isHeartbeat);
-        LOG.info("{}: appendEntries* reply {}", getMemberId(), toAppendEntriesReplyString(reply));
-        followerState.ifPresent(fs -> fs.updateLastRpcTime(FollowerState.UpdateType.APPEND_COMPLETE));
-        return CompletableFuture.completedFuture(reply);
-      }
+        if (!proto.getInitializing() && lifeCycle.compareAndTransition(State.STARTING, State.RUNNING)) {
+          role.startFollowerState(this, Op.APPEND_ENTRIES);
+        }
+        followerState = updateLastRpcTime(FollowerState.UpdateType.APPEND_START);
+
+        // Check that the append entries are not inconsistent. There are 3
+        // scenarios which can result in inconsistency:
+        //      1. There is a snapshot installation in progress
+        //      2. There is an overlap between the snapshot index and the entries
+        //      3. There is a gap between the local log and the entries
+        // In any of these scenarios, we should return an INCONSISTENCY reply
+        // back to leader so that the leader can update this follower's next index.
+        final long inconsistencyReplyNextIndex = checkInconsistentAppendEntries(previous, entries);
+        if (inconsistencyReplyNextIndex > RaftLog.INVALID_LOG_INDEX) {
+          final AppendEntriesReplyProto reply = toAppendEntriesReplyProto(
+              leaderId, getMemberId(), currentTerm, followerCommit, inconsistencyReplyNextIndex,
+              AppendResult.INCONSISTENCY, callId, RaftLog.INVALID_LOG_INDEX, isHeartbeat);
+          LOG.info("{}: appendEntries* reply {}", getMemberId(), toAppendEntriesReplyString(reply));
+          followerState.ifPresent(fs -> fs.updateLastRpcTime(FollowerState.UpdateType.APPEND_COMPLETE));
+          return CompletableFuture.completedFuture(reply);
+        }
 
-      state.updateConfiguration(entries);
+        state.updateConfiguration(entries);
+      }
+    } finally {
+      future.ifPresent(CompletableFuture::join);
```

Comment:
I recommend using the wait mechanism to release the lock when stopping followers, candidates and leaders to avoid deadlock problems. Putting join at the end here may bring new concurrency problems

Reply:
ok, I'll write two version of code for comparison